include/aws/common/system_info.h

```diff
@@ -21,8 +22,48 @@ struct aws_cpu_info {
     bool suspected_hyper_thread;
 };
 
+struct aws_system_environment;
+
 AWS_EXTERN_C_BEGIN
 
+/**
+ * Allocates and initializes information about the system the current process is executing on.
+ * If successful returns an instance of aws_system_environment. If it fails, it will return NULL.
+ */
+AWS_COMMON_API
+struct aws_system_environment *aws_system_environment_load(struct aws_allocator *allocator);
```

Comment:
Does this design scale? Where you load all the info up front and returned cached results later?

That means, for every new getter you might want to add, the user MUST wait during load() for all N bits of info to be fetched, even if they weren't interested in all of them.

An alternative design is to lazily fetch info when a getter is called, and cache it. This would probably mean a lock around each function implementation.

I guess we could keep this API, load everything upfront for now, and change it later to the lazy fetching approach

Reply:
It scales better than needing to load, allocate, and deal with the complexity of mutexes and global state. I don't think those are ever the correct answer unless there's literally no other way to do it.